File: cli/CIBERSORTx.R
#!/usr/bin/env -S Rscript --no-save --quiet
source("renv/activate.R")
library(argparser)

# parse args
parser <- arg_parser("Deconvolute Samples")
parser <- add_argument(
  parser, "--dir",
  short = "-d",
  help = "path to run directory",
  default = ""
)
parser <- add_argument(
  parser, "--ref",
  short = "-r",
  help = "path to CIBERSORTx GEP",
  default = ""
)
parser <- add_argument(
  parser,
  "--id",
  short = "-i",
  help = "ID to use for outputs"
)
parser <- add_argument(
  parser,
  "--cores",
  short = "-c",
  help = "number of cores to use",
  default = 8
)
parser <- add_argument(
  parser,
  "--verbose",
  short = "-v",
  help = "should messages be printed? One of: DEBUG, INFO, WARN, ERROR",
  default = "INFO"
)
parser <- add_argument(
  parser,
  "--use-slurm",
  short = "-S",
  help = "should slurm be used for parallelization?",
  flag = TRUE
)
argv <- parse_args(parser)

suppressPackageStartupMessages({
  library(here)
  library(log4r)
})

logger <- logger(argv$verbose)

use_singularity <- FALSE
use_singularity <- tryCatch(
  {
    system("docker")
  },
  warning = function(w) {
    TRUE
  }
)

if (!is.logical(use_singularity)) use_singularity <- FALSE

debug(logger, paste0("Docker Tested with this final value for use_singularity: ", use_singularity))

if (argv$dir == "") {
  output_path <- paste0("outs/", argv$id)
} else {
  output_path <- paste0(parser$run_dir, "/outs/", argv$id)
}
if (!dir.exists(here(output_path))) {
  fatal(logger, "Output directory does not exist")
}

if (!file.exists(here(output_path, "cibersort_ref_input.txt"))) {
  info(logger, "CIBERSORTx Prepared files not detected.")

  sys_cmd <- paste(
    "./cli/make_CIBERSORT_ref.R",
    "-i", argv$id,
    "-d", argv$dir,
    "-v", argv$verbose
  )
  system(sys_cmd)
}

debug(logger, "Copying Data")
files <- list.files(here("cibersort_in"), full.names = TRUE)
file.copy(files, here(output_path))

info(logger, "Preparing commands for CIBERSORTx")


if (use_singularity) {
  info(logger, "Docker is not installed, falling back to singularity.")
  debug(logger, "Testing singularity")
  tryCatch(
    {
      system("singularity")
    },
    warning = function(w) {
      error(logger, "Singularity is not installed")
    }
  )
  cmd_start <- paste(
    "singularity run --pwd /src",
    "-B", paste0(here(output_path), ":/src/data"),
    "-B", paste0(here(output_path, "cibersort_results"), ":/src/outdir"),
    "docker://cibersortx/fractions"
  )
} else {
  info(logger, "Docker detected.")
  cmd_start <- paste(
    "docker run",
    "-v", paste0(here(output_path, "cibersort_in"), ":/src/data"),
    "-v", paste0(here(output_path, "cibersort_results"), ":/src/outdir"),
    "cibersortx/fractions"
  )
}

debug(logger, paste0("Using the username: ", Sys.getenv("EMAIL")))
debug(logger, paste0("Using the token: ", Sys.getenv("TOKEN")))

base_cmd <- paste(
  cmd_start,
  "--username", Sys.getenv("EMAIL"),
  "--verbose", "FALSE",
  "--token", Sys.getenv("TOKEN"),
  "--single_cell", "TRUE",
  "--outdir", here(output_path, "cibersort_results")
)

if (argv$verbose != "DEBUG") {
  debug_redir <- "1> /dev/null"
} else {
  debug_redir <- ""
}

sys_cmd <- paste(
  base_cmd,
  "--refsample", "cibersort_ref_input.txt",
  debug_redir
)
ref_filename <- "CIBERSORTx_cibersort_ref_input_inferred_refsample.txt"

if (!file.exists(here(output_path, "cibersort_results", ref_filename))) {
  info(logger, "Making CIBERSORTx Reference")
  system(sys_cmd)
} else {
  info(logger, "Reference Found. Skipping Reference Creation")
}

file.copy(here(output_path, "cibersort_results", ref_filename), here(output_path, "cibersort_in"))

info(logger, "Running CIBERSORTx")

sys_cmd <- function(mixture) {
  paste(
    base_cmd,
    "--sigmatrix", ref_filename,
    "--mixture", mixture,
    debug_redir
  )
}

if (argv$use_slurm) {
  if (!require("slurmR")) {
    slurm_install <- tryCatch(devtools::install_github("USCbiostats/slurmR"),
      error = function(e) {
        info(logger, "slurmR not installed. Not using slurm")
        NULL
      }
    )
    if (is.null(slurm_install)) {
      argv$use_slurm <- FALSE
    }
  }
}

if (argv$use_slurm) {
  lapply_cus <- function(X, FUN) {
    Slurm_lapply(X, FUN, plan = "collect", njobs = length(X))
  }
} else {
  lapply_cus <- function(X, FUN) lapply(X = X, FUN = FUN)
}

files_list <- list("beat_aml_1_sam_test.txt")

cibersort_results <- lapply_cus(X = files_list, FUN = function(x) {
  system(sys_cmd(x))
})

source(here("lib/WriteInvocation.R"))
WriteInvocation(argv, output_path = here(output_path, "invocation"))

File: cli/CIBERSORTx.py
#!/usr/bin/env python3

import subprocess
import argparse
import os
import shutil
import logging
import sys
import uuid
from datetime import datetime
import lib

parser = argparse.ArgumentParser(description = "Deconvolute Samples", add_help=False)

runtime = parser.add_argument_group("Runtime Arguments")
runtime.add_argument("-d", "--dir", help="path to run directory", default=None)
runtime.add_argument("-i", "--id", help="ID to use for outputs", required=True)
runtime.add_argument("-m", "--mixture", help="what mixture file to use", required=True, nargs="+")

flags = parser.add_argument_group("optional argumuments")
flags.add_argument("-h", "--help", help = "show this help message and exit", action = "help")
flags.add_argument("-v", "--verbose", help="should debug messages be printed", action="count", default=0)
flags.add_argument("-D", "--debug", help="Should stdout be printed from CIBERSORTx", action="store_true")
flags.add_argument("--batch-correct", help="Should the integrated data from seurat be used?", action="store_true")
flags.add_argument("--batch-correct-method", "-M", help="which method from Seurat should be used?", choices=["CCA", "RPCA"])

group = flags.add_mutually_exclusive_group()
group.add_argument("-X", "--B-mode", help="Should B mode batch correction be applied?", action="store_true")
group.add_argument("-S", "--S-mode", help="Should S mode batch correction be applied?", action="store_true")

argv = parser.parse_args()
verbose_level = 2 - argv.verbose
logging.basicConfig(level=verbose_level, format="%(levelname)s [%(asctime)s] %(name)s %(message)s", datefmt="%Y-%m-%d %H:%M:%S")

if (len(argv.mixture) == 1):
    name_to_use = argv.mixture[0]
else:
    name_to_use = lib.concatenate_list_data(argv.mixture)

logger = logging.getLogger("CIBERSORTx: [{}]".format(name_to_use))

if (argv.batch_correct): 
    if (argv.batch_correct_method is None):
        logger.critical("Batch correction method not set")
        sys.exit(1)

docker_singularity_cmd = "docker"
try: 
    subprocess.run("docker", stderr=subprocess.DEVNULL)
except FileNotFoundError as e:
    logger.info("Docker not Found. Falling back to Singularity")
    docker_singularity_cmd = "singularity"

if docker_singularity_cmd == "singularity":
    try: 
        subprocess.run("singularity", stderr=subprocess.DEVNULL)
    except FileNotFoundError as e:
        logger.critical("Singularity Not Found")
        sys.exit(1)

if docker_singularity_cmd == "singularity":
    use_singularity = True
    logger.info("Using Singularity")
elif docker_singularity_cmd == "docker":
    use_singularity = False
    logger.info("Using Docker")
else:
    logger.critical("Only Singularity and Docker are supported")
    sys.exit(1)

# set up outputs
if argv.dir == None:
    output_path = "outs/" + argv.id
    plots_path = "plots/" + argv.id
else:
    output_path = "{}/outs/{}".format(argv.dir, argv.id)
    plots_path = "{}/plots/{}".format(argv.dir, argv.id)

random_string = uuid.uuid1()
cibersort_output = lib.setup_results_dir(argv)
name_of_output_directory = "{}/{}".format(cibersort_output, random_string)

logger.debug("Output Directory: {}".format(name_of_output_directory))

try:
    os.mkdir("{}/{}".format(output_path, cibersort_output))
except FileExistsError as e:
    logger.debug("Output directory already created")

try: 
    os.mkdir("{}/{}".format(output_path, name_of_output_directory))
except FileExistsError as e:
    files = os.listdir("{}/{}".format(output_path, name_of_output_directory))
    for f in files:
        if f.startswith("temp"):
            logger.warning("Temporary files already in temporary directory. Two CIBERSORTx instances may share an run directory.")
    logger.debug("Temporary directory already exists")

# set up bind mounts
if docker_singularity_cmd == "docker":
    path_to_run_folder = os.path.abspath(os.getcwd())
    output_path = "{}/{}".format(path_to_run_folder, output_path)
input_bind = output_path + ":/src/data"
output_bind = "{}/{}:/src/outdir".format(output_path, name_of_output_directory)
   
user_email = os.getenv("EMAIL")
user_token = os.getenv("TOKEN")

if user_email is None or user_token is None:
    logger.critical("Enviorment Variables not found!")
    logger.debug("User Email: {}".format(user_email))
    logger.debug("User Token: {}".format(user_token))
    raise RuntimeError("Env vars not found!")

logger.info("Preparing Run Commands")
if use_singularity:
    container_cmd = ["singularity", "run", "--pwd", "/src", "-B", input_bind, "-B", output_bind, "docker://cibersortx/fractions"]
else:
    container_cmd = ["docker", "run", "-v", input_bind, "-v", output_bind, "cibersortx/fractions"]

logger.debug("Testing if Reference can be found")

bc_ext = "no_bc"
if argv.batch_correct:
    bc_ext = argv.batch_correct_method

ref_filename = "CIBERSORTx_cell_type_sourceGEP.txt"
cibersort_cmd = ["--username", user_email, "--verbose", str(argv.debug).upper(), "--token", user_token, " --single_cell", "TRUE", "--outdir", "{}/{}".format(output_path, name_of_output_directory)]

if os.path.isfile("{}/{}".format(output_path, ref_filename)):
    logger.info("Existing Reference Found")
elif os.path.isfile("{}/{}_cibersort_results/{}".format(output_path, bc_ext, ref_filename)):
    logger.info("Existing Reference Found")
    logger.info("Copying Reference to CIBERSORTx input directory")
    source_path = "{}/{}_cibersort_results/{}".format(output_path, bc_ext, ref_filename)
    destination_path = "{}/{}".format(output_path, ref_filename)
    shutil.copyfile(source_path, destination_path)
else:
    logger.critical("Reference Not Found. Please run CIBERSORTx_ref.py")
    sys.exit(1)
    
def run_cmd(mixture):
    specific_args = ["--sigmatrix", ref_filename, "--mixture", mixture, "--label", mixture]
    bc_args = ["--rmbatchBmode", str(argv.B_mode).upper(), "--rmbatchSmode", str(argv.S_mode).upper(), "--refsample", "cibersort_ref_input.txt"]
    res = [container_cmd, cibersort_cmd, specific_args, bc_args]
    flattened = [val for sublist in res for val in sublist]
    return flattened

for dat in argv.mixture:
    logger.info("Running CIBERSORTx on {}".format(dat))
    source_path = "{}/{}".format("cibersort_in", dat)
    destination_path = "{}/{}".format(output_path, dat)
    shutil.copyfile(source_path, destination_path)
    logger.debug("The run command is: {}".format(" ".join(run_cmd(dat))))
    if argv.debug:
        subprocess_fun = None
    else:
        subprocess_fun = subprocess.DEVNULL
    start_time = datetime.now()
    subprocess.run(" ".join(run_cmd(dat)), shell=True, stdout=subprocess_fun)
    end_time = datetime.now()
    elapsed_time = end_time - start_time
    logger.info("CIBERSORTx completed in {}".format(elapsed_time))

files = os.listdir("{}/{}/{}".format(output_path, name_of_output_directory, random_string))

remove = True
for f in files:
    source_path = "{}/{}/{}".format(output_path, name_of_output_directory, f)
    destination_path = "{}/{}/{}".format(output_path, cibersort_output, f)
    try: 
        shutil.copyfile(source_path, destination_path)
    except IsADirectoryError as e:
        logger.error("Error moving files. Temporary Directory will not be deleted!")
        remove = False

if remove:
    shutil.rmtree(source_path)
 
lib.write_invocation(argv, output_path)
